mava/components/jax/building/guardrails.py

```diff
@@ -0,0 +1,277 @@
+    ):
+        """Save the config"""
+        self.config = config
+
+    def on_building_init_start(self, builder: SystemBuilder) -> None:
+        """Check that all component requirements are satisfied"""
+        for component in builder.callbacks:
+            for required_component in component.required_components():
+                required_component_present = False
+
+                for system_component in builder.callbacks:
+                    if isinstance(system_component, required_component):
+                        required_component_present = True
+                        break
+
+                if not required_component_present:
+                    raise Exception(
+                        f"""
+                        Component {component} requires other components
+                        {component.required_components()}
+                        to be present in the system.
+                        Component {required_component} is not in the system.
+                        """
+                    )
+
+    @staticmethod
+    def name() -> str:
+        """Static method that returns component name."""
+        return "component_dependency_guardrails"
+
+    @staticmethod
+    def config_class() -> Optional[Callable]:
+        """Optional class which specifies the dataclass/config object for the component.
+
+        Returns:
+            config class/dataclass for component.
+        """
+        return ComponentDependencyGuardrailsConfig
+
+    @staticmethod
+    def required_components() -> List[Type[Callback]]:
+        """List of other Components required in the system for this Component to function.
+
+        None required.
+
+        Returns:
+            List of required component classes.
+        """
+        return []
+
+
+@dataclass
+class ComponentDependencyDebuggerConfig:
+    show_full_component_path: bool = False
+    fig_width_ratios = [1, 5]
+    fig_inch_width: float = 18.5
+    fig_inch_height: float = 10.5
+    component_name_fontsize: int = 12
+    node_label_font_size: int = 8
+    required_node_color: str = "#ADD8E6"
+    leaf_required_node_color: str = "#90EE90"
+    detached_node_color: str = "#F0F0F0"
+    dependent_node_color: str = "#D3D3D3"
+    cycle_edge_color: str = "#FF0000"
+    default_edge_color: str = "#000000"
+    layout_scale: int = 2
+    node_size_increment: int = 80
+    min_node_size: int = 190
+    dependency_graph_save_path: str = "./component_dependency_map.png"
+
+
+class ComponentDependencyDebugger(Component):
+    def __init__(
+        self,
+        config: ComponentDependencyDebuggerConfig,
+    ):
+        """Save the config"""
+        self.config = config
+
+    def on_building_init_end(self, builder: SystemBuilder) -> None:  # noqa: C901
+        """Compute component dependencies and save graph."""
+        components: List = builder.callbacks[:]
+
+        # Add required components to list
+        additional_required_components = []
```

Comment:
The naming is a bit difficult to follow. Maybe something like `all_required_components`? Otherwise it seems like there are additional components required in the system

Reply:
I realise now that we are actually checking for additional components. Could you please give and example? Becuase it seems like all the components should be included in the list created on line 102? Otherwise, shouldn't this cause an error because we are requiring components that aren't in the system?

Reply:
Agreed, will rename to `all_required_components`, it was named badly by me

Reply:
> 

Just following up re this question for incase it got lost in the mix.